### PR TITLE
Add proper func tests

### PR DESCRIPTION
**Pull request by Deezzir**
Adds the proper functional tests for the snap. Includes:

- Tests for the Prometheus endpoint
- Tests for all current configs

### CODE REVIEW

File: tests/functional/test_snap.py [hunk 1]
```diff
@@ -1,5 +1,153 @@
...
+    """Get the data from an endpoint."""
+    response = urllib.request.urlopen(endpoint)
+    return response.read().decode("utf-8")
+
+
+@retry(wait=wait_fixed(2), stop=stop_after_delay(10))
+def _check_bind_config(bind: str) -> None:
+    """Check if a service is listening on a specific bind."""
+    assert 0 == subprocess.call(
+        f"nc -z localhost {bind.lstrip(':')}".split()
```
Comment (aieri) [inline tests/functional/test_snap.py:49]: suggestion: while within the context of these tests this is adequate, it's not technically telling you if the snap is the one listening on that port. You could consider using lsof for a more precise check. Something like `lsof -sTCP:LISTEN -i tcp:<bind port>` should give you the process that is listening on that specific port. Further output filtering switches may be available.

**Review by aieri — DISMISSED** [review]
Nice! And I see the influence of dcgm snap tests :)
I have only added a non-blocking suggestion.

File: tests/functional/test_snap.py [hunk 2]
```diff
@@ -1,5 +1,157 @@
...
+        f"sudo snap unset {SNAP_NAME} {config_parent}.{config}".split()
+    ), f"Failed to unset {config}"
+
+
+def _get_start_cmd(process: str) -> str:
+    return subprocess.check_output(f"ps -C {process} -o cmd -ww".split(), text=True)
+
+
+def _check_config(config_parent: str, config: str):
+    """Check if a configuration exists in the snap configuration."""
```
Comment (samuelallan72) [inline tests/functional/test_snap.py:74]: Just a suggestion on wording to make it clearer.

```suggestion
def _assert_config_exists(config_parent: str, config: str):
    """Check if a configuration exists in the snap configuration."""
```

File: tests/functional/test_snap.py [hunk 3]
```diff
@@ -1,5 +1,157 @@
...
+    result = subprocess.check_output(f"sudo snap get {SNAP_NAME} -d".split(), text=True)
+    smartctl_snap_config = json.loads(result.strip())
+    assert (
+        config in smartctl_snap_config[config_parent]
+    ), f"{config} is not in the snap configuration"
+
+
+@contextmanager
+def _config_setup(config_str, new_value):
+    """Set up a context manager to test snap configuration."""
```
Comment (samuelallan72) [inline tests/functional/test_snap.py:84]: This could have better docs and naming, to include that it's a context manager so you can use it to test the snap with custom config, and have the config reverted afterwards even if the test fails.

I like this pattern - it makes the tests really clear and concise. :+1:

File: tests/functional/test_snap.py [hunk 4]
```diff
@@ -1,5 +1,157 @@
...
+@retry(wait=wait_fixed(2), stop=stop_after_delay(10))
+def _check_service_active() -> None:
+    """Check if a systemd service is in a activemstate."""
+    assert 0 == subprocess.call(
+        f"sudo systemctl is-active --quiet {SYSTEMD_SERVICE}".split()
+    ), f"{SYSTEMD_SERVICE} is not running"
+
+
+@retry(wait=wait_fixed(5), stop=stop_after_delay(30))
+def _check_endpoint(endpoint: str) -> None:
```
Comment (samuelallan72) [inline tests/functional/test_snap.py:30]: This suggestion is probably too wordy, but something a bit more descriptive would be helpful:

```suggestion
def _assert_endpoint_returns_success_response(endpoint: str) -> None:
```

**Review by samuelallan72 — CHANGES_REQUESTED** [review]
The tests look really nice! I just flagged the tox.ini change, and left some suggestions for naming to make the tests clearer. :)

File: tox.ini [hunk 5]
```diff
@@ -28,6 +28,7 @@ deps =
     {[testenv:func]deps}
 
 [testenv:reformat]
+envdir = {toxworkdir}/lint
```
[the remark this answers is not in the document]
Comment (chanchiwai-ray) [reply tox.ini:31]: Please also see https://tox.wiki/en/latest/upgrading.html#reuse-of-environments

**Review by aieri — APPROVED** [review]
